File: runtime/typescript/src/Int.ts
import { Codec } from "./Codec.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";
import { encodeVLQ, decodeVLQ } from "./VLQ.js";

export type Int = bigint;

export const int: Codec<Int> = {
    read(reader: FormatReader): Promise<Int> {
        return decodeVLQ(reader, true);
    },

    write(writer: FormatWriter, value: Int): Promise<void> {
        return encodeVLQ(writer, true, value);
    },
};

File: runtime/java/src/main/java/dev/argon/verilization/runtime/Int.java
package dev.argon.verilization.runtime;

import java.math.BigInteger;
import java.io.IOException;

public final class Int {
    private Int() {}

    public static BigInteger fromInteger(int i) {
        return BigInteger.valueOf(i);
    }

    public static BigInteger fromInteger(long l) {
        return BigInteger.valueOf(l);
    }

    public static BigInteger fromInteger(BigInteger i) {
        return i;
    }

    public static final Codec<BigInteger> intCodec = new Codec<BigInteger>() {
        @Override
        public BigInteger read(FormatReader reader) throws IOException {
            return VLQ.decodeVLQ(reader, true);
        }

        @Override
        public void write(FormatWriter writer, BigInteger value) throws IOException {
            VLQ.encodeVLQ(writer, true, value);
        }

    };
}
